File: assessment-3/defrag-map.py
import sys

def process_map(lines):
    if not lines:
        return
    width = len(lines[0])
    sizes = [line.count('*') for line in lines]
    
    # Place files from the right (fastest) in priority order
    result = [['.' for _ in range(width)] for _ in lines]
    pos = width  # next free position (from the right)
    
    for i, size in enumerate(sizes):
        if size == 0:
            continue
        pos -= size
        for j in range(pos, pos + size):
            result[i][j] = '*'
    
    for row in result:
        print(''.join(row))

def main():
    maps = []
    current = []
    
    for line in sys.stdin:
        line = line.rstrip('\n')
        if line == '':
            if current:
                maps.append(current)
                current = []
        else:
            current.append(line)
    
    if current:
        maps.append(current)
    
    for i, m in enumerate(maps):
        process_map(m)
        if i < len(maps) - 1:
            print()

if __name__ == "__main__":
    main()

File: assessment-3/defrag-map.rb
def process_map(lines)
  return if lines.empty?

  width = lines[0].length
  sizes = lines.map { |line| line.count('*') }

  result = Array.new(lines.size) { Array.new(width, '.') }
  pos = width

  sizes.each_with_index do |size, i|
    next if size == 0
    pos -= size
    (pos...pos + size).each do |j|
      result[i][j] = '*'
    end
  end

  result.each { |row| puts row.join }
end

maps = []
current = []

STDIN.each_line do |line|
  line = line.chomp
  if line.empty?
    if !current.empty?
      maps << current
      current = []
    end
  else
    current << line
  end
end

maps << current unless current.empty?

maps.each_with_index do |m, i|
  process_map(m)
  puts if i < maps.size - 1
end